### PR TITLE
Degridifying unpacked phi to break degeneracies in angular separations for packed candidates

### PR DESCRIPTION
**Pull request by ferencek**
This PR tries to address the issue first reported (as far as I am aware) in https://github.com/vhbb/cmssw/pull/60 and discussed in https://hypernews.cern.ch/HyperNews/CMS/get/jet-algorithms/360.html and https://hypernews.cern.ch/HyperNews/CMS/get/physTools/3352.html

The code changes introduced here should be safe for backporting to 74X since they do not affect the MiniAOD production, only its unpacking.

@gpetruc @arizzi @rappoccio @gkasieczka  @slava77

### CODE REVIEW

**Comment by arizzi** [comment]
I'd rather fix that in the calling code (JetAlgorithm) rather than here. Btw, any reason to smear phi rather than eta? did we check which of the two is more degenerate?

**Comment by ferencek** [comment]
FastJet works with rapidity and phi so from its point of view, phi should be more degenerate. Also, just changing phi ensures that the total energy of the particle does not change.

File: DataFormats/PatCandidates/src/PackedCandidate.cc [hunk 1]
```diff
@@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
 }
 
 void pat::PackedCandidate::unpack() const {
-    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
+    float pt = MiniFloatConverter::float16to32(packedPt_);
+    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
+    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
```
Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: where is the matching -shift in ::pack?

Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: When I suggested to move the degridification to packing, I meant to do it so that after one pack->unpack cycle, additional pack->unpack keeps unpacked values unchanged.

Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: ... I'm not sure if that matters though

Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: By construction, the degridification cannot be done at the packing stage, since it is the packing that introduces the grid (the packing process involves mapping to integers so there are no fractional integers available). The degridification can only be introduced at the unpacking stage.

File: DataFormats/PatCandidates/src/PackedCandidate.cc [hunk 2]
```diff
@@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
 }
 
 void pat::PackedCandidate::unpack() const {
-    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
+    float pt = MiniFloatConverter::float16to32(packedPt_);
+    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
```
Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: Btw, how about replacing `shift` with `sign*shift` where `sign` would be something like
```c++
double sign = ( ( int(pt*10) % 2 == 0 ) ? 1 : -1 );
```
This should be random enough.

File: DataFormats/PatCandidates/src/PackedCandidate.cc (hunk 1 above)
Comment (arizzi) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: at constr we do pack() + unpack() to be sure that transient packed already
contains the approx.
did you check if this op is stable?
Il 11/giu/2015 19:06, "Dinko Ferencek" <notifications@github.com> ha
scritto:

Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: I haven't checked explicitly but this should be the case since the phi shifts are within 0.1 of the packed integer values.

Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: I see `p4_` is used in `packVtx()` which means that the packed vertex information would change if packing is done after unpacking. However, that seems to be the case already now and as far as I can tell, that's why `packBoth()` was introduced. Please correct me if I'm misinterpreting something.

Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: On 6/11/15 12:05 PM, Dinko Ferencek wrote:
> In DataFormats/PatCandidates/src/PackedCandidate.cc
> <https://github.com/cms-sw/cmssw/pull/9554#discussion_r32243557>:
> 
>> @@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
>>  }
>>  
>>  void pat::PackedCandidate::unpack() const {
>> -    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
>> +    float pt = MiniFloatConverter::float16to32(packedPt_);
>> +    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
>> +    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
> 
> By construction, the degridification cannot be done at the packing
> stage, since it is the packing that introduces the grid (the packing
> process involves mapping to integers so there are no fractional integers
> available). The degridification can only be introduced at the unpacking
> stage.


As I suggested earlier on the HN, packing of a non-linear f(x) to
integers f_i
and then unpacking to x = fInverse (f_i) will give x_i = fInverse(f_i)
values not on the regular grid.

This approach is stable under pack->unpack cycle for all the x_i values.

For eta, it could be packing of theta (in email I mentioned eta**1.00001)

For phi, it could be packing of cos(phi) [+ 1 bit for pi offset]

Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: @slava, even if we were packing in theta, the grid would still exist in eta. Granted, it would not be uniform in eta, but degeneracies would still be possible. Furthermore, even if at the same time we had a non-linear packing of phi, there would be some symmetry among the 4 quadrants in the eta-phi plane and degeneracies would in principle still be possible (although I would assume they would be significantly less likely than now).

I don't think it makes sense to pack in cos(phi) since the measured phi resolution should be more or less invariant under phi. Hence, packing in phi is what makes most sense. Concerning packing in theta, yes, it probably makes more sense from the experimental point of view. However, constructors of 4-vector classes generally do not take theta as input which means that packing and unpacking would involve computationally more intensive non-linear functions. So here again I think it makes more sense to pack eta directly. In fact, the grid in eta becomes in some sense too dense in the forward region so there is no real issue.

Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: OK, I tried to suggest something stable under pack unpack operations.
Smoother transformation could always be picked up.

The feedback on this PR so far sounds like the PR is a no go anyways.

Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: Where is the degeneracy, actually?
Is it at the distance corresponding to the jet distance parameter (do we
have the largest as 1.2?),
or is it across the board?

If we use 0.1*pt, will it still work for the heavy flavor ID using
ghosts with very low momentum?


On 6/11/15 10:18 PM, Dinko Ferencek wrote:
> In DataFormats/PatCandidates/src/PackedCandidate.cc
> <https://github.com/cms-sw/cmssw/pull/9554#discussion_r32288004>:
> 
>> @@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
>>  }
>>  
>>  void pat::PackedCandidate::unpack() const {
>> -    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
>> +    float pt = MiniFloatConverter::float16to32(packedPt_);
>> +    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
>> +    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
> 
> @slava <https://github.com/slava>, even if we were packing in theta, the
> grid would still exist in eta. Granted, it would not be uniform in eta,
> but degeneracies would still be possible. Furthermore, even if at the
> same time we had a non-linear packing of phi, there would be some
> symmetry among the 4 quadrants in the eta-phi plane and degeneracies
> would in principle still be possible (although I would assume they would
> be significantly less likely than now).
> 
> I don't think it makes sense to pack in cos(phi) since the measured phi
> resolution should be more or less invariant under phi. Hence, packing in
> phi is what makes most sense. Concerning packing in theta, yes, it
> probably makes more sense from the experimental point of view. However,
> constructors of 4-vector classes generally do not take theta as input
> which means that packing and unpacking would involve computationally
> more intensive non-linear functions. So here again I think it makes more
> sense to pack eta directly. In fact, the grid in eta becomes in some
> sense too dense in the forward region so there is no real issue.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9554/files#r32288004>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------

Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: Since particles are placed on an eta-phi grid, the degeneracy occurs in dR separation of particle pairs. In other words, there can be more than one pair with the same dR separation and this is particularly troublesome for the CA algorithm which uses dR as the distance measure (anti-kt and kt break the degenracy by also using particle pt). These degeneracies mostly affect early stages of jet clustering so essentially jets of any size are affected, not just fat jets with large cone size.

The ghost association will still work just fine since the particle pt's are not changed (0.1*pt, for pt<1, only determines by how much to shift the phi value in units of the grid spacing).

Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: About the ghosts, my question was mostly about how their phi is defined:
if we use packedPhi_/2^15 + 0.1*pt/2^15
if pt is ~1E-19 (or what's the value?), then it will be below double
precision and ghosts will be on the grid.


On 6/12/15 8:13 AM, Dinko Ferencek wrote:
> In DataFormats/PatCandidates/src/PackedCandidate.cc
> <https://github.com/cms-sw/cmssw/pull/9554#discussion_r32314079>:
> 
>> @@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
>>  }
>>  
>>  void pat::PackedCandidate::unpack() const {
>> -    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
>> +    float pt = MiniFloatConverter::float16to32(packedPt_);
>> +    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
>> +    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
> 
> Since particles are placed on an eta-phi grid, the degeneracy occurs in
> dR separation of particle pairs. In other words, there can be more than
> one pair with the same dR separation and this is particularly
> troublesome for the CA algorithm which uses dR as the distance measure
> (anti-kt and kt break the degenracy by also using particle pt). These
> degeneracies mostly affect early stages of jet clustering so essentially
> jets of any size are affected, not just fat jets with large cone size.
> 
> The ghost association will still work just fine since the particle pt's
> are not changed (0.1*pt, for pt<1, only determines by how much to shift
> the phi value in units of the grid spacing).
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9554/files#r32314079>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------

Comment (ferencek) [inline DataFormats/PatCandidates/src/PackedCandidate.cc]: Indeed, this could lead to a ghost association that is not uniquely defined, but again, only for the CA algorithm. However, ghosts are always (or at least in the code I am aware of) handled separately so the would not undergo this kind of shifting procedure using their "ghost" pt.

**Comment by ferencek** [comment]
Here are some additional check I promised to perform. The first set of plots shown the difference in the jet multiplicity between AOD and MiniAOD for AK8 and CA8 jets before and after the phi degridification fix.

* **AK8 jet multiplicity before the fix:**
h2_njetdiff_ak8_before

* **AK8 jet multiplicity after the fix:**
h2_njetdiff_ak8_after

* **CA8 jet multiplicity before the fix:**
h2_njetdiff_ca8_before

* **CA8 jet multiplicity after the fix:**
h2_njetdiff_ca8_after

Overall, the fix has no major impact on the MiniAOD vs AOD differences. Also, CA jets show larger MiniAOD vs AOD differences, as can be expected.

The second set of plots shows the relative jet pt difference between the original and reclustered Nth jet in MiniAOD before and after the fix.

* **Relative jet pt difference for AK8 jets before the fix:**
h2_jetptreldiffminiaod_ak8_before

* **Relative jet pt difference for AK8 jets after the fix:**
h2_jetptreldiffminiaod_ak8_after

* **Relative jet pt difference for CA8 jets before the fix:**
h2_jetptreldiffminiaod_ca8_before

* **Relative jet pt difference for CA8 jets after the fix:**
h2_jetptreldiffminiaod_ca8_after

As expected, AK jets are very stable under reclustering even without the fix. With the fix, both AK and CA jets match exactly the original jets.

The third set of plots shows the relative jet pt difference between the Nth jet in MiniAOD and AOD before and after the fix.

* **Relative jet pt difference for AK8 jets before the fix:**
h2_jetptreldiff_ak8_before

* **Relative jet pt difference for AK8 jets after the fix:**
h2_jetptreldiff_ak8_after

* **Relative jet pt difference for CA8 jets before the fix:**
h2_jetptreldiff_ca8_before

* **Relative jet pt difference for CA8 jets after the fix:**
h2_jetptreldiff_ca8_after

Here again, AK jets are more stable when going from AOD to MiniAOD. Overall, the fix has no major impact on the MiniAOD vs AOD differences.

I also explicitly checked and the introduced phi shifts do not spoil the stability of the candidate momentum upon multiple pack and unpack cycles.

In summary, this PR should be complete at this point. We still need to wait for a response from FastJet authors on a possible fix inside the FastJet itself. In case they are no willing to provide a fix or think it is not necessary, this PR is a possible alternative solution.

**Comment by arizzi** [comment]
so you see an effect on AK too?
is there a pt cut? is the cut on uncorrected pt for both?

**Comment by ferencek** [comment]
@arizzi, are you referring to MiniAOD vs AOD differences or reclustered vs original in MiniAOD?

Concerning the pt cut, no, there is no pt cut applied and all pt's are raw pt's that you get by clustering PF candidates.

**Comment by arizzi** [comment]
I mean MiniAOD vs AOD...

**Comment by ferencek** [comment]
Thank you for clarifying. Yes, even AK jets have some MiniAOD vs AOD differences but that is expected to happen to some degree I would say. In any case, the effect is significantly smaller than for CA jets.

It is also interesting that the reclustered CA jets in MiniAOD differ from the original jets in MiniAOD as much as the original jets in MiniAOD differ from the AOD jets. This just illustrates the point you were making about CA jets in the email thread on HyperNews.

**Comment by slava77** [comment]
Hi Dinko,

What's happening in the vertical overflow bin?
210/212 before/after for AK8
3641/3625 before/after for CA8

**Comment by ferencek** [comment]
Hi Slava,

That was just a stupid mistake in my code where I was subtracting unsigned ints. I should have realized earlier that the sharp cut off at 0 was not physical. This has now been fixed (https://github.com/ferencek/cmssw/commit/f378129172cf2e7f7465b6f13dd18cd9d718118f) and the updated, correct plots are attached below.

* AK8 jet multiplicity before the fix:
h2_njetdiff_ak8_before

* AK8 jet multiplicity after the fix:
h2_njetdiff_ak8_after

* CA8 jet multiplicity before the fix:
h2_njetdiff_ca8_before

* CA8 jet multiplicity after the fix:
h2_njetdiff_ca8_after

I also forgot to mention that the plots were produced using all events in /RelValTTbar_13/CMSSW_7_4_1-PU50ns_MCRUN2_74_V8_gensim71X-v1/GEN-SIM-RECO

**Comment by arizzi** [comment]
ah ok, now I'm less worried for AK jets (btw can you make it in log or add
the test to understand how much stuff is in the purple tails of AK) ?

I think this proves that CA is unstable, to clarify what I mean we should
try to compare AK vs AK and CA vs CA on two version of tracker alignment
and see what happens.

**Comment by ferencek** [comment]
@arizzi, here are the AK plots in log scale.

* **AK8 jet multiplicity before the fix:**
h2_njetdiff_ak8_before

* **AK8 jet multiplicity after the fix:**
h2_njetdiff_ak8_after

**Comment by ferencek** [comment]
Now it is more than 1 week that we haven't heard back from the FastJet authors. @rappoccio, did they maybe reply to you only?

Given that the pseudo-random phi shifts (at most 1/10 of the phi grid spacing and decreasing with pt) do not introduce additional bias wrt AOD, I think this PR should be safe to include.

Can we at least trigger tests to confirm that MiniAOD production is not affected, just its reading.

**Comment by rappoccio** [comment]
I did not hear from them, no. I suggest that we do something ourselves.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ferencek** [comment]
Tests were successful so I suggest moving forward with this one.

**Comment by monttj** [comment]
+1
It seems we all agreed with this modification.

**Comment by ferencek** [comment]
The last check I wanted to show is the comparison of jet pt and CSVv2IVF distributions before and after the degridification fix. All distributions are made from /RelValTTbar_13/CMSSW_7_4_1-PU50ns_MCRUN2_74_V8_gensim71X-v1/MINIAODSIM

* **Jet pt spectrum** (solid red: from slimmedJets stored in MiniAOD, dashed blue: AK4CHS jets reclustered from MiniAOD, dotted green: AK4CHS jets reclustered from MiniAOD after introducing the fix) jet_pt
jet_pt

Lines are basically on top of each other so everything looks fine here.

* **Jet CSVv2IVF discriminator distribution** (solid red: from slimmedJets stored in MiniAOD, dashed blue: from AK4CHS jets reclustered from MiniAOD, dotted green: from AK4CHS jets reclustered from MiniAOD after introducing the fix) jet_csvv2ivf
jet_csvv2ivf

Here, as expected, there are some differences between the CSVv2IVF discriminators derived from AOD and MiniAOD (note that discriminators stored in MiniAOD are actually derived from AOD). However, what is important is that the blue and green lines are on top of each other so the fix does not spoil the b-tagging performance.

I actually wanted to post this here (instead of under #9809) to have all the checks in one place. Sorry for the duplication.

**Comment by ferencek** [comment]
Here is a reply from the FastJet authors, who essentially suggested to follow the approach implemented in this PR.

-------- Forwarded Message --------
Subject: Re: [Fastjet] Possible bug in FastJet affecting jet reclustering
Date: Thu, 25 Jun 2015 14:06:10 +0200
From: Matteo Cacciari <matteo.cacciari@cern.ch>
To: Salvatore Rappoccio <Salvatore.Rappoccio@cern.ch>, Dinko Ferencek <ferencek@physics.rutgers.edu>
CC: Email the developers (subscription not required) <fastjet@projects.hepforge.org>

Hi Sal,

Here's the long overdue reply, apologies for the delay.

As a rule, degeneracies are almost impossible to treat predictably: the issue
is that different machine architectures can treat the rounding differently and
so while you might have

  c - b = b - a

on one machine (e.g. x86_64), while on another (e.g. i386) you might end up with

  c - b > b - a

So even if one has a special treatment that makes FJ behave “predictably”,
breaking the degeneracy say as if c-b<b-a when it detects c-b=b-a, you could
always end up running on a machine that determines that c - b > b - a and so
bypasses that special treatment for degeneracies.

One should be aware that the degeneracy in clustering distance for grid-spaced
rap,phi values is there not just for C/A, but for all standard recombination
algs, including anti-kt. However for anti-kt, the degeneracy in initial
clustering choice nearly always turns out to be irrelevant when it comes to
the composition of the final jets. In contrast, C/A is much more sensitive.

The only workaround we see is to manually break the degeneracy. There is a FAQ
entry on the FastJet web pages, but it was been written years ago and
admittedly it is not as clear as it could be.

Random shifts are in a way the safest option (one can store a seed for the
shifts in order to guarantee reproducibility), because they shouldn’t
introduce any bias, at least on average.

Another alternative is to deform rapidity-azimuth space, e.g.

    phi -> phi + epsilon phi^2
    rap -> rap + epsilon (rap- rapmin)^2

In our own internal analysis software, this is an option we have for our
calorimeter class, precisely to be able to work around degeneracies. Of course
it might introduce a systematic (e.g. favouring biasing jets to in one
rapidity direction versus another), and as we actually have never used that
option much, we never studied the systematics in detail.

There are other potential workarounds, e.g. introducing small powers of
ptmin,ptmax, into the clustering distance. This is considerably more work,
especially to get fast implementations, and there is again no guarantee that
you would avoid biases (though you would probably not get phi or
rapidity-dependent biases at least).

Regarding using the pt to add shifts in rap,phi, yes, that too can work.

Sorry not to be able to give you a more helpful reply…

Cheers,
Matteo, Gregory & Gavin

**Comment by ferencek** [comment]
Is there a plan to include this PR before 750 is closed? I see it was forward ported to 76X and so far no problems were reported. Also note that a backport PR (#10008) was created and should be safe to merge since it only impacts the reading of MiniAOD, not its production.

This PR would be very useful for any analysis-level use cases where jet clustering and re-clustering is performed, such as JetToolbox.

**Comment by davidlange6** [comment]
+1
